### ISSUE TITLE
QDF mode non-friendly to non-ASCII text

### ISSUE THREAD

**Issue by jidanni** [qpdf/qpdf#337]
QDF mode is for editing PDF.
It makes
`  /Author <feff5ed657f976db>`
when pdfinfo file.pdf makes
Author:         廖培盛

QDF mode is for editing PDF.
It makes some kind binary blob,
when pdftotext file.pdf - makes
歸化國籍有殊勳於我國者認定原則...
(I.e., correct UTF-8)

Seen on
https://www.ris.gov.tw/documents/data/2/4/4e050f2b-9778-4a7d-a764-241e43d659f1.pdf
https://www.ris.gov.tw/documents/data/2/4/004639a9-52d3-4a1e-b542-3627d2b4ab31.pdf

**Comment by jberkenbilt** [comment]
I believe you are seeing expected, normal behavior that is not reflective of any incorrect handling of non-ASCII text. A tool like pdftotext extracts text from a PDF file that way PDF represents the text and presents it in a more normal way. In PDF, text strings are encoded in one of two different ways. For metadata, such as /Author, text containing characters that are not representable with PDFDocEncoding (basically ISO-Latin-1 plus a few additional characters all encoded as a single byte per character) are encoded using big-ending UTF-16. What you see in <feff5ed...> is the UTF-16 encoded string represented as a sequence of bytes. You can see this in a terminal:

```
% echo -n 廖培盛| iconv -t utf16be | od -t x1
0000000 5e d6 57 f9 76 db
0000006
```

There you see the same byte sequence of 5ed657f976db as in the PDF fie. The feff in the PDF is simply a marker that indicates that the string is UTF-16 encoded.

For text that appears in page contents, it's more complex. In that case, characters are mapped from arbitrary codepoint values, usually in order of appearance in the file, on a per-font basis, and the characters you actually see in PDF content streams don't look anything like the characters. This is frequently true even for English text, but it is always true for text containing non-ASCII characters that can't be represented in a single-byte encoding system supported by the PDF spec.

I hope this explanation helps. If you still think there is a problem, please feel free to reopen the issue.

**Comment by jidanni** [comment]
I'm just saying that you need a new "UDF" mode,
where CJK users can edit a CJK file with the same comfort as ASCII users can currently edit a pure ASCII file.

So with the new mode all the decoding work is done before presenting it to the user, and then it is all encoded back when the user hands it back, via the "fix" method...

**Comment by jberkenbilt** [comment]
Okay, thanks. Unfortunately, this is just not something that I am planning on doing. qpdf can help you work with the raw PDF file, but it's never going to be a tool that helps you edit files in a text/font-aware fashion. It's too far away from any functionality that exists in qpdf or from any of qpdf's goals. Also note that mostly qdf files are not suitable for editing content in arbitrary files containing ASCII text because of character maps and subsetted fonts. But I appreciate the suggestion.

**Comment by jidanni** [comment]
Yes, it seems the user at best could edit within a single line safely.
E.g., changing some words. That's life within the PDF.
